### PR TITLE
Add `delete_patterns` option to `upload_folder`

### PR DESCRIPTION
**Pull request by Wauplin**
Step 1 for #1352: add a `delete_patterns` parameter to `upload_folder`. It makes sense in a lot of cases to delete remote files before uploading new ones.

The workflow is:
1. if `delete_patterns` is provided, list remote files (can be optimized once https://github.com/huggingface/moon-landing/pull/5580 is deployed)
2. apply the pattern filter => create a list of CommitOperationDelete (never delete ".gitattributes")
3. if a file is both deleted and overwritten in the same commit, drop the deletion operation
4. (not done) we could optimize this in case the user deletes an entire folder. At the moment it's 1 file = 1 delete operation. Seems error-prone to try to decode the patterns manually so let's do it later if we see it's a limitation.

Included in PR:
- refactored `_prepare_upload_folder_additions` (fixed a bug) + implement `_prepare_upload_folder_deletions`
- add mocked tests for `upload_folder` to check the CommitOperation list is correct
- adapted ModelHubMixin + keras/pytorch/fastai integrations
- adapted upload docs + some docstrings

Step 2 for #1352 will be adding the possibility for automatic "chained commits" in `create_commit` using a PR if the list of operations is too big (see description there).

### CODE REVIEW

File: docs/source/guides/upload.mdx [hunk 1]
```diff
@@ -56,24 +56,54 @@ Specify the path of the file to upload, where you want to upload the file to in
...
+```
 
 Use the `allow_patterns` and `ignore_patterns` arguments to specify which files to upload. These parameters accept either a single pattern or a list of patterns.
 Patterns are Standard Wildcards (globbing patterns) as documented [here](https://tldp.org/LDP/GNU-Linux-Tools-Summary/html/x11655.htm).
 If both `allow_patterns` and `ignore_patterns` are provided, both constraints apply. By default, all files from the folder are uploaded.
 
 ```py
->>> from huggingface_hub import HfApi
->>> api = HfApi()
 >>> api.upload_folder(
```
Comment (LysandreJik) [inline docs/source/guides/upload.mdx:81]: (nit) Depending on how you test this (or if you use doctests at all, which I don't think is the case), you may need to still redeclare the previous if you want to isolate each statement. Still clean this way so no need to change.

File: src/huggingface_hub/hf_api.py [hunk 2]
```diff
@@ -2620,21 +2620,25 @@ def upload_folder(
         parent_commit: Optional[str] = None,
         allow_patterns: Optional[Union[List[str], str]] = None,
         ignore_patterns: Optional[Union[List[str], str]] = None,
+        delete_patterns: Optional[Union[List[str], str]] = None,
     ):
         """
-        Upload a local folder to the given repo. The upload is done
-        through a HTTP requests, and doesn't require git or git-lfs to be
-        installed.
+        Upload a local folder to the given repo. The upload is done through a HTTP requests, and doesn't require git or
```
Comment (LysandreJik) [inline src/huggingface_hub/hf_api.py:2626]:
```suggestion
        Upload a local folder to the given repo. The upload is done through a HTTP request and doesn't require git or
```

File: src/huggingface_hub/hf_api.py [hunk 3]
```diff
@@ -2620,21 +2620,25 @@ def upload_folder(
...
-        Use the `allow_patterns` and `ignore_patterns` arguments to specify which files
-        to upload. These parameters accept either a single pattern or a list of
-        patterns. Patterns are Standard Wildcards (globbing patterns) as documented
-        [here](https://tldp.org/LDP/GNU-Linux-Tools-Summary/html/x11655.htm). If both
-        `allow_patterns` and `ignore_patterns` are provided, both constraints apply. By
-        default, all files from the folder are uploaded.
+        Use the `delete_patterns` argument to specify remote files you want to delete. Input type is the same as for
+        `allow_patterns` (see above). If `path_in_repo` is also provided, the patterns are matched against paths
+        relative to this folder. For example, `upload_folder(..., path_in_repo="experiment", delete_patterns="logs/*")`
+        will delete any remote file under `./experiment/logs/`. Note that the `.gitattributes` file will not be deleted
```
Comment (LysandreJik) [inline src/huggingface_hub/hf_api.py:2640]:
```suggestion
        will delete any remote file under `experiment/logs/`. Note that the `.gitattributes` file will not be deleted
```

File: src/huggingface_hub/hf_api.py [hunk 4]
```diff
@@ -2714,6 +2722,19 @@ def upload_folder(
...
+        # Useful if you don't know exactly which log files have already being pushed
+        >>> upload_folder(
+        ...     folder_path="local/checkpoints",
+        ...     path_in_repo="remote/experiment/checkpoints",
+        ...     repo_id="username/my-dataset",
+        ...     repo_type="datasets",
+        ...     token="my_token",
+        ...     delete_patterns="**/logs/*.txt",
+        ... )
+        # "https://huggingface.co/datasets/username/my-dataset/tree/main/remote/experiment/checkpoints"
```
Comment (LysandreJik) [inline src/huggingface_hub/hf_api.py:2735]: If this is the return of the method above, I don't think you need to precede it with `#`. I thought this was a comment for clarity but I believe it's actually the return of the method?

**Review by LysandreJik — APPROVED** [review]
Looks great to me!

File: src/huggingface_hub/hf_api.py (hunk 4 above)
Comment (Wauplin) [reply src/huggingface_hub/hf_api.py:2735]:
```suggestion
        "https://huggingface.co/datasets/username/my-dataset/tree/main/remote/experiment/checkpoints"
```

True, thanks for noticing

**Comment by Wauplin** [comment]
Thanks for the careful review :+1: 
I have addressed the comments and will now merge it :)

**Pull request merged**